firedrake/cython/dmswarm.pyx

```diff
@@ -0,0 +1,504 @@
+# cython: language_level=3
+
+# Utility functions to derive global and local numbering from DMSwarm
+import cython
+import numpy as np
+from firedrake.petsc import PETSc
+from mpi4py import MPI
+from pyop2.datatypes import IntType
+from libc.string cimport memset
+from libc.stdlib cimport qsort
+cimport numpy as np
+cimport mpi4py.MPI as MPI
+cimport petsc4py.PETSc as PETSc
+
+np.import_array()
+
+include "petschdr.pxi"
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def mark_entity_classes(PETSc.DM swarm):
+    """Mark all points in a given DMSwarm according to the PyOP2 entity
+    classes:
+
+    core   : owned and not in send halo
+    owned  : owned and in send halo
+    ghost  : in halo
+
+    ..note:: this is a copy and paste job from the equivalent function
+             in dmplex.pyx with the relevant modifications made for a
+             DMSwarm.
+
+    :arg swarm: The DMSwarm object encapsulating the mesh topology
+    """
+    cdef:
+        PetscInt pStart, pEnd, cStart, cEnd
+        PetscInt c, ci, p
+        PetscInt nleaves
+        const PetscInt *ilocal = NULL
+        PetscBool non_exec
+        const PetscSFNode *iremote = NULL
+        PETSc.SF point_sf = None
+        PetscBool is_ghost, is_owned
+        DMLabel lbl_core, lbl_owned, lbl_ghost
+
+    pStart = 0
+    pEnd = swarm.getLocalSize()
+
+    swarm.createLabel("pyop2_core")
+    swarm.createLabel("pyop2_owned")
+    swarm.createLabel("pyop2_ghost")
+
+    CHKERR(DMGetLabel(swarm.dm, b"pyop2_core", &lbl_core))
+    CHKERR(DMGetLabel(swarm.dm, b"pyop2_owned", &lbl_owned))
+    CHKERR(DMGetLabel(swarm.dm, b"pyop2_ghost", &lbl_ghost))
+
+    if swarm.comm.size > 1:
+        # Mark ghosts from point overlap SF
+        point_sf = swarm.getPointSF()
+        CHKERR(PetscSFGetGraph(point_sf.sf, NULL, &nleaves, &ilocal, NULL))
+        for p in range(nleaves):
+            CHKERR(DMLabelSetValue(lbl_ghost, ilocal[p], 1))
+    else:
+        # If sequential mark all points as core
+        for p in range(pStart, pEnd):
+            CHKERR(DMLabelSetValue(lbl_core, p, 1))
+        return
+
+    CHKERR(DMLabelCreateIndex(lbl_ghost, pStart, pEnd))
+
+    # Mark all remaining points as core
+    CHKERR(DMLabelCreateIndex(lbl_owned, pStart, pEnd))
+    for p in range(pStart, pEnd):
+        CHKERR(DMLabelHasPoint(lbl_owned, p, &is_owned))
+        CHKERR(DMLabelHasPoint(lbl_ghost, p, &is_ghost))
+        if not is_ghost and not is_owned:
+            CHKERR(DMLabelSetValue(lbl_core, p, 1))
+    CHKERR(DMLabelDestroyIndex(lbl_owned))
+    CHKERR(DMLabelDestroyIndex(lbl_ghost))
```

Comment:
OK, so this function should then be simplified significantly to remove most of the references to `is_owned` (maybe add a comment at the top that due to lack of connectivity points are either core or ghost). You probably still need to make the (now empty) `pyop2_owned` label.

I would do:

```
if comm.size == 1:
    for p in range(pStart, pEnd):
        DMLabelSetValue(core, p, 1)
else:
    getgraph(...)
    for p in range(nleaves):
        DMLabelSetValue(ghost, ilocal[p], 1)
    # Mark remaining points as core
    DMLabelCreateIndex(ghost, ...)
    for p in range(pStart, pEnd):
        DMLabelHasPoint(ghost, ...)
        if not is_ghost:
            DMLabelSetValue(core, p, 1)
    DMLabelDestroyIndex(ghost, ...)
```

Reply:
I believe that this has been superseded by the discussion above since we don't have a pointSF set up for the DMSwarm.